File: app/src/main/java/com/liferlighdow/device/MainActivity.java
package com.liferlighdow.device;

import android.app.Activity;
import android.graphics.Color;
import android.os.Build;
import android.os.Bundle;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.TextView;

public class MainActivity extends Activity {

    private FrameLayout contentFrame;
    private LinearLayout bottomNav;
    private HardwareActivity hardwareActivity;
    private ApplicationsActivity appsActivity;
    private TemperatureActivity tempActivity;
    private SettingsActivity settingsActivity;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        View rootLayout = findViewById(R.id.root_layout);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT_WATCH) {
            rootLayout.setOnApplyWindowInsetsListener((v, insets) -> {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                    v.setPadding(insets.getInsets(android.view.WindowInsets.Type.systemBars()).left,
                            insets.getInsets(android.view.WindowInsets.Type.systemBars()).top,
                            insets.getInsets(android.view.WindowInsets.Type.systemBars()).right,
                            insets.getInsets(android.view.WindowInsets.Type.systemBars()).bottom);
                } else {
                    v.setPadding(insets.getSystemWindowInsetLeft(),
                            insets.getSystemWindowInsetTop(),
                            insets.getSystemWindowInsetRight(),
                            insets.getSystemWindowInsetBottom());
                }
                return insets;
            });
        }

        contentFrame = findViewById(R.id.content_frame);
        bottomNav = findViewById(R.id.bottom_nav);

        hardwareActivity = new HardwareActivity(this);
        appsActivity = new ApplicationsActivity(this);
        tempActivity = new TemperatureActivity(this);
        settingsActivity = new SettingsActivity(this, new SettingsActivity.OnSettingsChangedListener() {
            @Override
            public void onThemeChanged() {
                applyTheme();
            }

            @Override
            public void onUnitChanged() {
                // Temperature values will refresh on next 5s cycle
            }

            @Override
            public void onQuickTestRequested(String type) {
                if ("SCREEN".equals(type)) {
                    runScreenTest();
                }
            }
        });

        findViewById(R.id.nav_hardware).setOnClickListener(v -> showHardware());
        findViewById(R.id.nav_apps).setOnClickListener(v -> showApps());
        findViewById(R.id.nav_temp).setOnClickListener(v -> showTemp());
        findViewById(R.id.nav_settings).setOnClickListener(v -> showSettings());

        applyTheme();
        showHardware();
    }

    private void applyTheme() {
        int bgColor = ThemeManager.getBackgroundColor(this);
        int textColor = ThemeManager.getTextColor(this);
        int navBg = ThemeManager.isLightMode(this) ? Color.parseColor("#F5F5F5") : Color.BLACK;

        contentFrame.setBackgroundColor(bgColor);
        bottomNav.setBackgroundColor(navBg);

        for (int i = 0; i < bottomNav.getChildCount(); i++) {
            View v = bottomNav.getChildAt(i);
            if (v instanceof TextView) {
                TextView tv = (TextView) v;
                tv.setTextColor(textColor);
                // Refresh drawables if needed, but they are white which works okay on dark, 
                // on light we might want them black.
                if (ThemeManager.isLightMode(this)) {
                    tv.getCompoundDrawables()[1].setColorFilter(Color.BLACK, android.graphics.PorterDuff.Mode.SRC_IN);
                } else {
                    tv.getCompoundDrawables()[1].clearColorFilter();
                }
            }
        }

        hardwareActivity.applyTheme();
        appsActivity.applyTheme();
        tempActivity.applyTheme();
        settingsActivity.refreshUI();
    }

    private void showHardware() {
        tempActivity.stopUpdates();
        contentFrame.removeAllViews();
        contentFrame.addView(hardwareActivity.getView());
    }

    private void showApps() {
        tempActivity.stopUpdates();
        contentFrame.removeAllViews();
        contentFrame.addView(appsActivity.getView());
    }

    private void showTemp() {
        contentFrame.removeAllViews();
        contentFrame.addView(tempActivity.getView());
        tempActivity.startUpdates();
    }

    private void showSettings() {
        tempActivity.stopUpdates();
        bottomNav.setVisibility(View.VISIBLE);
        contentFrame.removeAllViews();
        contentFrame.addView(settingsActivity.getView());
    }

    private int screenTestStep = 0;
    private void runScreenTest() {
        tempActivity.stopUpdates();
        bottomNav.setVisibility(View.GONE);
        contentFrame.removeAllViews();
        
        View testView = new View(this);
        int[] colors = {Color.RED, Color.GREEN, Color.BLUE, Color.WHITE, Color.BLACK};
        testView.setBackgroundColor(colors[0]);
        screenTestStep = 0;

        testView.setOnClickListener(v -> {
            screenTestStep++;
            if (screenTestStep < colors.length) {
                v.setBackgroundColor(colors[screenTestStep]);
            } else {
                bottomNav.setVisibility(View.VISIBLE);
                showSettings(); // Exit test
                android.widget.Toast.makeText(this, "Screen test finished", android.widget.Toast.LENGTH_SHORT).show();
            }
        });

        contentFrame.addView(testView);
        
        // Use a simple full screen flag if available
        if (Build.VERSION.SDK_INT >= 19) {
            testView.setSystemUiVisibility(
                View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
                | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                | View.SYSTEM_UI_FLAG_FULLSCREEN);
        }

        android.widget.Toast.makeText(this, "Tap to cycle colors", android.widget.Toast.LENGTH_SHORT).show();
    }

    @Override
    protected void onPause() {
        super.onPause();
        tempActivity.stopUpdates();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        tempActivity.stopUpdates();
    }
}

File: app/src/main/java/com/liferlighdow/device/MainActivity.kt
package com.liferlighdow.device

import android.graphics.Color
import android.graphics.PorterDuff
import android.os.Build
import android.os.Bundle
import android.view.View
import android.view.WindowInsets
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.activity.viewModels
import com.liferlighdow.device.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    private val hardwareViewModel: HardwareViewModel by viewModels()
    
    private lateinit var hardwareActivity: HardwareActivity
    private lateinit var appsActivity: ApplicationsActivity
    private lateinit var appDetailActivity: AppDetailActivity
    private lateinit var tempActivity: TemperatureActivity
    private lateinit var settingsActivity: SettingsActivity
    
    private var isDetailView = false
    private var lastBackPressTime: Long = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        // Handle Window Insets for edge-to-edge
        binding.root.setOnApplyWindowInsetsListener { v, insets ->
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                val systemBars = insets.getInsets(WindowInsets.Type.systemBars())
                v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            } else {
                @Suppress("DEPRECATION")
                v.setPadding(insets.systemWindowInsetLeft, insets.systemWindowInsetTop, insets.systemWindowInsetRight, insets.systemWindowInsetBottom)
            }
            insets
        }

        hardwareActivity = HardwareActivity(this, hardwareViewModel)
        appsActivity = ApplicationsActivity(this)
        appDetailActivity = AppDetailActivity(this)
        tempActivity = TemperatureActivity(this)
        settingsActivity = SettingsActivity(this, object : SettingsActivity.OnSettingsChangedListener {
            override fun onThemeChanged() {
                applyTheme()
            }

            override fun onUnitChanged() {
                // Refresh logic if needed
            }

            override fun onQuickTestRequested(type: String) {
                if (type == "SCREEN") runScreenTest()
            }
        })

        binding.navHardware.setOnClickListener { showHardware() }
        binding.navApps.setOnClickListener { showApps() }
        binding.navTemp.setOnClickListener { showTemp() }
        binding.navSettings.setOnClickListener { showSettings() }

        applyTheme()
        showHardware()
    }

    private fun applyTheme() {
        val bgColor = ThemeManager.getBackgroundColor(this)
        val textColor = ThemeManager.getTextColor(this)
        val navBg = if (ThemeManager.isLightMode(this)) Color.parseColor("#F5F5F5") else Color.BLACK

        binding.contentFrame.setBackgroundColor(bgColor)
        binding.bottomNav.setBackgroundColor(navBg)

        for (i in 0 until binding.bottomNav.childCount) {
            val v = binding.bottomNav.getChildAt(i)
            if (v is android.widget.TextView) {
                v.setTextColor(textColor)
                if (ThemeManager.isLightMode(this)) {
                    v.compoundDrawables[1]?.setColorFilter(Color.BLACK, PorterDuff.Mode.SRC_IN)
                } else {
                    v.compoundDrawables[1]?.clearColorFilter()
                }
            }
        }

        hardwareActivity.applyTheme()
        appsActivity.applyTheme()
        tempActivity.applyTheme()
        settingsActivity.refreshUI()
    }

    fun showHardware() {
        isDetailView = false
        tempActivity.stopUpdates()
        binding.contentFrame.removeAllViews()
        binding.contentFrame.addView(hardwareActivity.getView())
    }

    fun showApps() {
        isDetailView = false
        tempActivity.stopUpdates()
        binding.contentFrame.removeAllViews()
        binding.contentFrame.addView(appsActivity.getView())
    }

    fun showAppDetail(app: ApplicationsActivity.AppEntry) {
        isDetailView = true
        tempActivity.stopUpdates()
        binding.contentFrame.removeAllViews()
        binding.contentFrame.addView(appDetailActivity.getView(app))
    }

    fun showTemp() {
        isDetailView = false
        binding.contentFrame.removeAllViews()
        binding.contentFrame.addView(tempActivity.getView())
        tempActivity.startUpdates()
    }

    fun showSettings() {
        isDetailView = false
        tempActivity.stopUpdates()
        binding.bottomNav.visibility = View.VISIBLE
        binding.contentFrame.removeAllViews()
        binding.contentFrame.addView(settingsActivity.getView())
    }

    @Deprecated("Deprecated in Java")
    override fun onBackPressed() {
        @Suppress("DEPRECATION")
        if (isDetailView) {
            showApps()
        } else {
            if (lastBackPressTime + 2000 > System.currentTimeMillis()) {
                super.onBackPressed()
            } else {
                Toast.makeText(this, "Press back again to exit", Toast.LENGTH_SHORT).show()
                lastBackPressTime = System.currentTimeMillis()
            }
        }
    }

    private var screenTestStep = 0
    private fun runScreenTest() {
        tempActivity.stopUpdates()
        binding.bottomNav.visibility = View.GONE
        binding.contentFrame.removeAllViews()
        
        val testView = View(this)
        val colors = intArrayOf(Color.RED, Color.GREEN, Color.BLUE, Color.WHITE, Color.BLACK)
        testView.setBackgroundColor(colors[0])
        screenTestStep = 0

        testView.setOnClickListener { v ->
            screenTestStep++
            if (screenTestStep < colors.size) {
                v.setBackgroundColor(colors[screenTestStep])
            } else {
                binding.bottomNav.visibility = View.VISIBLE
                showSettings()
                Toast.makeText(this, "Screen test finished", Toast.LENGTH_SHORT).show()
            }
        }

        binding.contentFrame.addView(testView)
        
        @Suppress("DEPRECATION")
        testView.systemUiVisibility = (View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
                or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                or View.SYSTEM_UI_FLAG_FULLSCREEN)

        Toast.makeText(this, "Tap to cycle colors", Toast.LENGTH_SHORT).show()
    }

    override fun onPause() {
        super.onPause()
        tempActivity.stopUpdates()
        hardwareActivity.onPause()
    }

    override fun onResume() {
        super.onResume()
        hardwareActivity.onResume()
    }

    override fun onDestroy() {
        super.onDestroy()
        tempActivity.stopUpdates()
        hardwareActivity.destroy()
    }
}
